lib/WeBWorK/FormatRenderedProblem.pm

```diff
@@ -0,0 +1,314 @@
+	}
+	my %cssFilesAdded;    # Used to avoid duplicates
+	my @extra_css_files;
+	for (@cssFiles) {
+		next if $cssFilesAdded{ $_->{file} };
+		$cssFilesAdded{ $_->{file} } = 1;
+		if ($_->{external}) {
+			push(@extra_css_files, $_);
+		} else {
+			push(@extra_css_files, { file => getAssetURL($formLanguage, $_->{file}), external => 0 });
+		}
+	}
+
+	# Third party JavaScript
+	# The second element of each array in the following is whether or not the file is a theme file.
+	# The third element is a hash containing the necessary attributes for the script tag.
+	my @third_party_js = map { [ getAssetURL($formLanguage, $_->[0]), $_->[1] ] } (
+		[ 'node_modules/jquery/dist/jquery.min.js',                            {} ],
+		[ 'node_modules/jquery-ui-dist/jquery-ui.min.js',                      {} ],
+		[ 'node_modules/iframe-resizer/js/iframeResizer.contentWindow.min.js', {} ],
+		[ "js/apps/MathJaxConfig/mathjax-config.js",                           { defer => undef } ],
+		[ 'node_modules/mathjax/es5/tex-svg.js',                               { defer => undef, id => 'MathJax-script' } ],
+		[ 'node_modules/bootstrap/dist/js/bootstrap.bundle.min.js',            { defer => undef } ],
+		[ "js/apps/Problem/problem.js",                                                { defer => undef } ],
+		[ "js/apps/Problem/submithelper.js",                                           { defer => undef } ],
+		[ "js/apps/CSSMessage/css-message.js",                                 { defer => undef } ],
+	);
+
+	# Get the requested format. (outputFormat or outputformat)
+	# override to static mode if showCorrectAnswers has been set
+	my $formatName = $inputs_ref->{showCorrectAnswers} && !$inputs_ref->{isInstructor}
+		? 'static'
+		: $inputs_ref->{outputFormat} // $inputs_ref->{outputformat} // 'default';
+
+	# Add JS files requested by problems via ADD_JS_FILE() in the PG file.
+	my @extra_js_files;
+	if (ref($rh_result->{flags}{extra_js_files}) eq 'ARRAY') {
+		my %jsFiles;
+		for (@{ $rh_result->{flags}{extra_js_files} }) {
+			next if $jsFiles{ $_->{file} };
+			$jsFiles{ $_->{file} } = 1;
+			my %attributes = ref($_->{attributes}) eq 'HASH' ? %{ $_->{attributes} } : ();
+			if ($_->{external}) {
+				push(@extra_js_files, $_);
+			} else {
+				push(@extra_js_files,
+					{ file => getAssetURL($formLanguage, $_->{file}), external => 0, attributes => $_->{attributes} });
+			}
+		}
+	}
+
+	# Set up the problem language and direction
+	# PG files can request their language and text direction be set.  If we do not have access to a default course
+	# language, fall back to the $formLanguage instead.
+	# TODO: support for right-to-left languages
+	my %PROBLEM_LANG_AND_DIR =
+		get_problem_lang_and_dir($rh_result->{flags}, 'auto:en:ltr', $formLanguage);
+	my $PROBLEM_LANG_AND_DIR = join(' ', map {qq{$_="$PROBLEM_LANG_AND_DIR{$_}"}} keys %PROBLEM_LANG_AND_DIR);
+
+	# is there a reason this doesn't use the same button IDs?
```

Comment:
The button IDs were something like WWsubmit, and I don't exactly recall the others. Are these IDs consistent across rendering methods in WW2? I thought the choice of button IDs for the renderer were aligned to WW2, but the RPC template IDs were not what I expected.

Reply:
I think by "ID" you mean "name", and that is what is confusing me.  The html input id is not received in the form, the input name is.  In any case, no there is no consistency on these names.  The html2xml and render_rpc endpoints use one set of names for these inputs, regular webwork2 homework sets another, and gateway tests use yet another.  Really, none of those need to have the same names because they are completely separate rendering mechanisms.  For the render_rpc and html2xml endpoints, those names are what have been used, and those utilizing those endpoints probably expect them to be named that.